plugins/inputs/prometheus_scraper/metrics_type_handler.go

```diff
@@ -0,0 +1,197 @@
+type mCache struct {
+	t *scrape.Target
+}
+
+func (m *mCache) Metadata(metricName string) (scrape.MetricMetadata, bool) {
+	return m.t.Metadata(metricName)
+}
+
+// Adapter to ScrapeManager to retrieve the cache by job and instance
+type metadataService interface {
+	Get(job, instance string) (metadataCache, error)
+}
+
+type metadataServiceImpl struct {
+	sm ScrapeManager
+}
+
+func (t *metadataServiceImpl) Get(job, instance string) (metadataCache, error) {
+	targetGroupMap := t.sm.TargetsAll()
+	targetGroup, ok := targetGroupMap[job]
+
+	if !ok {
+		//when the job is replaced in relabel_config, TargetsAll() still return the map with old job name as key
+		//so we need to go over all the targets to find the matching job name
+		targetGroup = nil
+	checkJobLoop:
+		for _, potentialTargetGroup := range targetGroupMap {
+			for _, target := range potentialTargetGroup {
+				if target.Labels().Get(model.JobLabel) == job {
+					targetGroup = potentialTargetGroup
+					break checkJobLoop
+				}
+			}
+		}
+
+		if targetGroup == nil {
+			return nil, errors.New("unable to find a target group with job=" + job)
+		}
+	}
+
+	// from the same targetGroup, instance is not going to be duplicated
+	for _, target := range targetGroup {
+		if target.Labels().Get(model.InstanceLabel) == instance {
+			return &mCache{target}, nil
+		}
+	}
+	return nil, errors.New("unable to find a target with job=" + job + ", and instance=" + instance)
+}
+
+type ScrapeManager interface {
+	TargetsAll() map[string][]*scrape.Target
+}
+
+type metricsTypeHandler struct {
+	mc *metadataServiceImpl
+}
+
+func NewMetricsTypeHandler() *metricsTypeHandler {
+	return &metricsTypeHandler{}
+}
+
+func (mth *metricsTypeHandler) SetScrapeManager(scrapeManager ScrapeManager) {
+	if scrapeManager != nil {
+		mth.mc = &metadataServiceImpl{sm: scrapeManager}
+	}
+}
+
+// Return JobName and Instance
+func GetScrapeTargetInfo(pmb PrometheusMetricBatch) (string, string, error) {
+
+	for _, pm := range pmb {
+		job, ok := pm.tags[model.JobLabel]
+		if !ok {
+			continue
+		}
+		instance, ok := pm.tags[model.InstanceLabel]
+		if !ok {
+			continue
+		}
+		return job, instance, nil
+	}
+	return "", "", errors.New("No Job and Instance Label found.")
+}
+
+func isInternalMetric(metricName string) bool {
+	//For each endpoint, Prometheus produces a set of internal metrics. See https://prometheus.io/docs/concepts/jobs_instances/
+	if metricName == "up" || strings.HasPrefix(metricName, "scrape_") {
+		return true
+	}
+	return false
+}
+
+// Decorate the Metrics with Metric Types
+func (mf *metricsTypeHandler) Handle(pmb PrometheusMetricBatch) (result PrometheusMetricBatch) {
+	// Filter out Summary, Histogram and untyped Metrics and adding logging
+	jobName, instanceId, err := GetScrapeTargetInfo(pmb)
+	if err != nil {
+		log.Printf("E! Failed to get Job Name and Instance ID from Prometheus metrics. \n")
+		return result
+	}
+
+	mc, err := mf.mc.Get(jobName, instanceId)
```

Comment:
The variable name "mc" is confusing here, the second "mc" can be renamed as ms which stand for metadataService

Reply:
Will rename it.